akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/HttpCookie.scala

```diff
@@ -100,23 +178,71 @@ final case class HttpCookie(
   /** Java API */
   def getExpires: Optional[jm.DateTime] = expires.map(_.asJava).asJava
+
+  def withName(name: String): HttpCookie = copy(name = name)
+  def withValue(value: String): HttpCookie = copy(value = value)
+  /** Scala API */
+  def withExpires(dateTime: DateTime): HttpCookie = copy(expires = Some(dateTime))
   /** Java API */
-  def withExpires(dateTime: jm.DateTime): headers.HttpCookie = copy(expires = Some(dateTime.asScala))
-  /** Java API */
-  def withDomain(domain: String): headers.HttpCookie = copy(domain = Some(domain))
-  /** Java API */
-  def withPath(path: String): headers.HttpCookie = copy(path = Some(path))
-  /** Java API */
-  def withMaxAge(maxAge: Long): headers.HttpCookie = copy(maxAge = Some(maxAge))
-  /** Java API */
-  def withSecure(secure: Boolean): headers.HttpCookie = copy(secure = secure)
-  /** Java API */
-  def withHttpOnly(httpOnly: Boolean): headers.HttpCookie = copy(httpOnly = httpOnly)
+  def withExpires(dateTime: jm.DateTime): HttpCookie = copy(expires = Some(dateTime.asScala))
+
+  def withDomain(domain: String): HttpCookie = copy(domain = Some(domain))
+  def withPath(path: String): HttpCookie = copy(path = Some(path))
+  def withMaxAge(maxAge: Long): HttpCookie = copy(maxAge = Some(maxAge))
+  def withSecure(secure: Boolean): HttpCookie = copy(secure = secure)
+  def withHttpOnly(httpOnly: Boolean): HttpCookie = copy(httpOnly = httpOnly)
+  def withSameSite(sameSite: jm.headers.SameSite): HttpCookie = copy(sameSite = Option(sameSite.asScala()))
+
+  /** Scala API */
+  def withSameSite(sameSite: Option[SameSite]) = copy(sameSite = sameSite)
   /** Java API */
-  def withExtension(extension: String): headers.HttpCookie = copy(extension = Some(extension))
+  def withSameSite(sameSite: Optional[jm.headers.SameSite]): HttpCookie = copy(sameSite = sameSite.asScala.map(_.asScala()))
+
+  def withExtension(extension: String): HttpCookie = copy(extension = Some(extension))
 }
 
 object HttpCookie {
+
+  def apply(
+    name:      String,
+    value:     String,
+    expires:   Option[DateTime] = None,
+    maxAge:    Option[Long]     = None,
+    domain:    Option[String]   = None,
+    path:      Option[String]   = None,
+    secure:    Boolean          = false,
+    httpOnly:  Boolean          = false,
+    extension: Option[String]   = None,
+    sameSite:  Option[SameSite] = None
+  ) = new HttpCookie(name, value, expires, maxAge, domain, path, secure, httpOnly, extension, sameSite)
+
+  @deprecated("for binary compatibility", since = "10.2.0")
+  def apply(
+    name:      String,
+    value:     String,
+    expires:   Option[DateTime],
+    maxAge:    Option[Long],
+    domain:    Option[String],
+    path:      Option[String],
+    secure:    Boolean,
+    httpOnly:  Boolean,
+    extension: Option[String]
+  ) = new HttpCookie(name, value, expires, maxAge, domain, path, secure, httpOnly, extension, None)
+
+  @deprecated("for binary compatibility", since = "10.2.0")
```

Comment:
Let's add better deprecation hints where the methods are still public and the hints likely to be shown.

```suggestion
  @deprecated("Pattern matching on HttpCookie is deprecated because of the big number of fields and potential future compatibility hazards. Please use other means to check the fields.", since = "10.2.0")
```